common/dynamicconfig/collection.go

```diff
@@ -81,10 +113,86 @@ var (
 
 // NewCollection creates a new collection
 func NewCollection(client Client, logger log.Logger) *Collection {
-	return &Collection{
-		client:   client,
-		logger:   logger,
-		errCount: -1,
+	c := &Collection{
+		client:        client,
+		logger:        logger,
+		errCount:      -1,
+		subscriptions: make(map[Key]map[int]any),
+	}
+	s := DynamicConfigSubscriptionCallback.Get(c)()
+	c.callbackPool = goro.NewAdaptivePool(clock.NewRealTimeSource(), s.MinWorkers, s.MaxWorkers, s.TargetDelay, s.ShrinkFactor)
+	if subcli, ok := client.(SubscribableClient); ok {
+		c.cancelClientSubscription = subcli.Subscribe(c.keysChanged)
+	} else {
+		c.poller.Go(c.pollForChanges)
+	}
+	return c
+}
+
+func (c *Collection) Stop() {
+	c.poller.Cancel()
+	c.poller.Wait()
+	if c.cancelClientSubscription != nil {
+		c.cancelClientSubscription()
+	}
+	c.subscriptionLock.Lock()
+	defer c.subscriptionLock.Unlock()
+	c.callbackPool.Stop()
+}
+
+// Implement pingable.Pingable
+func (c *Collection) GetPingChecks() []pingable.Check {
+	return []pingable.Check{
+		{
+			Name:    "dynamic config callbacks",
+			Timeout: 5 * time.Second,
+			Ping: func() []pingable.Pingable {
+				var wg sync.WaitGroup
+				wg.Add(1)
+				c.callbackPool.Do(wg.Done)
```

Comment:
This is for the deadlock detector. It's configurable: by default it dumps all goroutine stacks, but can be configured to additionally start failing health checks, or to log.Fatal

Reply:
I fear someone will accidently use the update callback function to do the actual work as well instead of saving the updates and kicking off a goroutine to act on the updates. This work could be arbitrarily long and they will hold up one of the workers in the pool. Ex: where someone uses dynamic configs to scale up/down some resource.

How does the alternative look where the updates are called in independent Goroutines? The benefit is the dynamic config subscriptions are not affected the behavior of the consumer of this system. The downside is we may leak goroutines if they block forever (a different problem though).